content/advent-2016/testing-distributed-systems-in-go.md

```diff
@@ -0,0 +1,185 @@
+
+Naïve system tests assume ideal conditions: the network is stable, machine failures are absent. Production distributed systems are a world apart-- machines fail, latencies wildly fluctuate, and [network faults are commonplace](http://queue.acm.org/detail.cfm?id=2655736). A distributed system must still work despite these difficulties. Modeling these failures is extra work but essential. Otherwise, very real bugs will only surface during catastrophic  disasters in production.
+
+
+### How etcd is tested
+
+As of December 8, 2016, etcd weighs in at 114,815 lines of Go code. This is only code that could be considered exclusive to etcd; vendored and auto-generated code doesn’t count. Over half the code base, 60,033 lines, is dedicated to tests and testing infrastructure. Since so much of etcd is dedicated to tests, there are many kinds of tests. The testing infrastructure for etcd includes unit tests, integration tests, migration tests, end-to-end tests, benchmarks, regression tests, soak tests, stress tests, and functional tests. 
+
+Unit test checks input and output of a single component within the package. For example, raft package passes a message to its message-step function and checks the response.
+
+Client and server interactions are tested in integration tests: start etcd server using unix sockets, send client requests, and check responses from the server.
+
+End-to-end test configures the whole system locally and simulate the real-world operations: set up 3-node etcd cluster with actual etcd binaries, and verify that etcd command line interface is working correctly. These teste are run locally, integrated with our CI systems.
+
+
+### etcd’s "distributed" functional tester
+
+[functional-tester](https://github.com/coreos/etcd/tree/master/tools/functional-tester) verifies the correct behavior of etcd under various system and network malfunctions. It sets up an etcd cluster under high pressure loads and continuously injects failures into the cluster. Then it expects the etcd cluster to recover within a short amount of time. This has been extremely helpful to find critical bugs, *before anyone else*.
+
+Here's the overview of functional-tester:
+
+![etcd-functional-tester](/postimages/advent-2016/etcd-functional-tester.png)
+
+
+
+etcd functional test suite has two components: etcd-agent and etcd-tester. etcd-agent runs on every test machines and etcd-tester runs on a single machine as a testing controller. etcd-tester controls all the etcd-agent to start etcd clusters and simulate various failure cases.
```

Comment:
`etcd functional test suite has two components: etcd-agent and etcd-tester. etcd-agent runs on every test machines, and etcd-tester runs on a single machine as a testing controller. etcd-agent controlls the state of test machine that it resides on. It can stop, start, and restart an etcd node and much more such as manipulating network configuration. etcd-tester controlls the flow of functional test. It can simulate various failure cases by controlling etcd-agents and verifies the correctness of etcd under those failures.`

I think we might want to be a bit more specific on the relationship of etcd-agent and etcd-tester  so that it will be more clear to user on how we use those.